File: rust_core/crates/indicators/tests/python_parity.py
from __future__ import annotations

import json
import math
import sys
from pathlib import Path

import numpy as np

REPO_ROOT = Path(__file__).resolve().parents[4]
sys.path.insert(0, str(REPO_ROOT))

from src.old.backtest_engine.core.indicator_cache import IndicatorCache  # noqa: E402


def make_candle(close: float) -> dict[str, float]:
    return {
        "open": close - 0.01,
        "high": close + 0.02,
        "low": close - 0.02,
        "close": close,
        "volume": 1.0,
    }


def series_to_list(series) -> list[float | None]:
    values = series.to_numpy(dtype="float64")
    out: list[float | None] = []
    for v in values:
        if not np.isfinite(v):
            out.append(None)
        else:
            out.append(float(v))
    return out


def ema_with_warmup(values: list[float], period: int) -> list[float | None]:
    if period <= 0:
        raise ValueError("EMA period must be > 0")
    if len(values) < period:
        return [None] * len(values)

    alpha = 2.0 / (period + 1.0)
    out: list[float | None] = [None] * len(values)
    sma = float(np.mean(values[:period]))
    out[period - 1] = sma
    prev = sma

    for i in range(period, len(values)):
        v = values[i]
        if not np.isfinite(v):
            out[i] = prev
            continue
        prev = alpha * v + (1.0 - alpha) * prev
        out[i] = prev

    return out


def expand_and_ffill(
    reduced: list[float | None], indices: list[int], length: int
) -> list[float | None]:
    out: list[float | None] = [None] * length
    for i, idx in enumerate(indices):
        if idx < length and i < len(reduced):
            out[idx] = reduced[i]

    last: float | None = None
    for i, value in enumerate(out):
        if value is not None and math.isfinite(value):
            last = value
        elif last is not None:
            out[i] = last
    return out


def zscore_rolling(values: list[float], window: int) -> list[float | None]:
    if window <= 0:
        raise ValueError("window must be > 0")
    out: list[float | None] = [None] * len(values)
    if len(values) < window:
        return out

    for i in range(window - 1, len(values)):
        window_vals = values[i - window + 1 : i + 1]
        if not all(np.isfinite(v) for v in window_vals):
            continue
        mean = float(np.mean(window_vals))
        denom = window - 1
        if denom <= 0:
            continue
        variance = float(
            np.sum([(v - mean) ** 2 for v in window_vals]) / denom
        )
        std = math.sqrt(variance)
        if std > 0.0 and math.isfinite(std):
            out[i] = (values[i] - mean) / std
        else:
            out[i] = 0.0
    return out


def main() -> None:
    closes = [1.00, 1.10, 1.05, 1.20, 1.15, 1.30]
    m1_candles = [make_candle(c) for c in closes]

    h1_c1 = make_candle(1.05)
    h1_c2 = make_candle(1.25)
    h1_candles = [h1_c1] * 3 + [h1_c2] * 3

    multi = {
        "M1": {"bid": m1_candles, "ask": m1_candles},
        "H1": {"bid": h1_candles, "ask": h1_candles},
    }

    cache = IndicatorCache(multi)

    ema = ema_with_warmup(closes, period=3)

    h1_new_idx = cache._stepwise_indices("H1", "bid")
    reduced = [h1_candles[i]["close"] for i in h1_new_idx]
    reduced_ema = ema_with_warmup(reduced, period=3)
    ema_step = expand_and_ffill(reduced_ema, h1_new_idx, len(h1_candles))
    bb_step_upper, bb_step_mid, bb_step_lower = cache.bollinger_stepwise(
        "H1", "bid", period=3, std_factor=2.0
    )
    kz_step = cache.kalman_zscore_stepwise(
        "H1", "bid", window=3, R=0.5, Q=0.1
    )

    atr = cache.atr("M1", "bid", period=3)
    bb_upper, bb_mid, bb_lower = cache.bollinger(
        "M1", "bid", period=3, std_factor=2.0
    )
    zscore_roll = zscore_rolling(closes, window=3)
    zscore_ema = cache.zscore(
        "M1", "bid", window=3, mean_source="ema", ema_period=2
    )
    kalman_zscore = cache.kalman_zscore(
        "M1", "bid", window=3, R=0.5, Q=0.1
    )
    garch_vol = cache.garch_volatility(
        "M1",
        "bid",
        alpha=0.1,
        beta=0.8,
        omega=None,
        use_log_returns=True,
        scale=100.0,
        min_periods=2,
        sigma_floor=1e-6,
    )
    kalman_garch_z = cache.kalman_garch_zscore(
        "M1",
        "bid",
        R=0.01,
        Q=1.0,
        alpha=0.1,
        beta=0.8,
        omega=None,
        use_log_returns=True,
        scale=100.0,
        min_periods=2,
        sigma_floor=1e-6,
    )

    garch_local = cache.garch_volatility_local(
        "M1",
        "bid",
        idx=5,
        lookback=4,
        alpha=0.1,
        beta=0.8,
        omega=None,
        use_log_returns=True,
        scale=100.0,
        min_periods=2,
        sigma_floor=1e-6,
    )
    kgz_local = cache.kalman_garch_zscore_local(
        "M1",
        "bid",
        idx=5,
        lookback=4,
        R=0.01,
        Q=1.0,
        alpha=0.1,
        beta=0.8,
        omega=None,
        use_log_returns=True,
        scale=100.0,
        min_periods=2,
        sigma_floor=1e-6,
    )

    kgz_value = None
    if isinstance(kgz_local, float) and math.isfinite(kgz_local):
        kgz_value = float(kgz_local)

    payload = {
        "ema": ema,
        "ema_stepwise": ema_step,
        "bollinger_stepwise": {
            "upper": series_to_list(bb_step_upper),
            "middle": series_to_list(bb_step_mid),
            "lower": series_to_list(bb_step_lower),
        },
        "kalman_zscore_stepwise": series_to_list(kz_step),
        "atr": series_to_list(atr),
        "bollinger": {
            "upper": series_to_list(bb_upper),
            "middle": series_to_list(bb_mid),
            "lower": series_to_list(bb_lower),
        },
        "zscore_rolling": zscore_roll,
        "zscore_ema": series_to_list(zscore_ema),
        "kalman_zscore": series_to_list(kalman_zscore),
        "garch_volatility": series_to_list(garch_vol),
        "kalman_garch_zscore": series_to_list(kalman_garch_z),
        "garch_volatility_local": series_to_list(garch_local),
        "kalman_garch_zscore_local": kgz_value,
    }

    json.dump(payload, sys.stdout)


if __name__ == "__main__":
    main()

File: rust_core/crates/indicators/tests/python_parity.rs
use std::path::PathBuf;
use std::process::Command;

use omega_indicators::{
    ATR, BollingerBands, EMA, GarchLocalParams, GarchVolatility, Indicator, KalmanGarchLocalParams,
    KalmanGarchZScore, KalmanZScore, MultiOutputIndicator, MultiTfIndicatorCache, ZScore,
    ZScoreMeanSource,
};
use omega_types::{Candle, PriceType, Timeframe};
use serde::Deserialize;

#[derive(Debug, Deserialize)]
struct BollingerOutput {
    upper: Vec<Option<f64>>,
    middle: Vec<Option<f64>>,
    lower: Vec<Option<f64>>,
}

#[derive(Debug, Deserialize)]
struct PythonParityOutput {
    ema: Vec<Option<f64>>,
    ema_stepwise: Vec<Option<f64>>,
    bollinger_stepwise: BollingerOutput,
    atr: Vec<Option<f64>>,
    bollinger: BollingerOutput,
    zscore_rolling: Vec<Option<f64>>,
    zscore_ema: Vec<Option<f64>>,
    kalman_zscore: Vec<Option<f64>>,
    garch_volatility: Vec<Option<f64>>,
    kalman_garch_zscore: Vec<Option<f64>>,
    kalman_zscore_stepwise: Vec<Option<f64>>,
    garch_volatility_local: Vec<Option<f64>>,
    kalman_garch_zscore_local: Option<f64>,
}

#[test]
fn test_python_parity_stepwise_and_local() {
    let output = run_python_parity();

    let closes = [1.00, 1.10, 1.05, 1.20, 1.15, 1.30];
    let mut primary_bid = Vec::new();
    for (i, close) in closes.iter().enumerate() {
        primary_bid.push(make_candle(i as i64 * 60, *close));
    }
    let primary_ask = primary_bid.clone();

    let h1_closes = [1.05, 1.25];
    let mut h1_bid = Vec::new();
    for (i, close) in h1_closes.iter().enumerate() {
        h1_bid.push(make_candle(i as i64 * 180, *close));
    }
    let h1_ask = h1_bid.clone();

    let mut cache = MultiTfIndicatorCache::new(
        Timeframe::M1,
        primary_bid.clone(),
        primary_ask,
        vec![(Timeframe::H1, h1_bid, h1_ask)],
    );

    let ema = EMA::new(3).compute(&primary_bid);
    assert_series_close("ema", &output.ema, &ema, 1e-10);

    let ema_step = cache.ema_stepwise(Timeframe::H1, PriceType::Bid, 3);
    assert_series_close("ema_stepwise", &output.ema_stepwise, &ema_step, 1e-10);

    let (bb_upper, bb_middle, bb_lower) =
        cache.bollinger_stepwise(Timeframe::H1, PriceType::Bid, 3, 2.0);
    assert_series_close(
        "bollinger_stepwise.upper",
        &output.bollinger_stepwise.upper,
        &bb_upper,
        1e-10,
    );
    assert_series_close(
        "bollinger_stepwise.middle",
        &output.bollinger_stepwise.middle,
        &bb_middle,
        1e-10,
    );
    assert_series_close(
        "bollinger_stepwise.lower",
        &output.bollinger_stepwise.lower,
        &bb_lower,
        1e-10,
    );

    let kz_step = cache.kalman_zscore_stepwise(Timeframe::H1, PriceType::Bid, 3, 0.5, 0.1);
    assert_series_close(
        "kalman_zscore_stepwise",
        &output.kalman_zscore_stepwise,
        &kz_step,
        1e-10,
    );

    let garch_params = GarchLocalParams {
        alpha: 0.1,
        beta: 0.8,
        omega: None,
        use_log_returns: true,
        scale: 100.0,
        min_periods: 2,
        sigma_floor: 1e-6,
    };
    let garch_local =
        cache.garch_volatility_local(Timeframe::M1, PriceType::Bid, 5, 4, garch_params);
    assert_series_close(
        "garch_volatility_local",
        &output.garch_volatility_local,
        &garch_local,
        1e-10,
    );

    let kgz_params = KalmanGarchLocalParams {
        r: 0.01,
        q: 1.0,
        alpha: 0.1,
        beta: 0.8,
        omega: None,
        use_log_returns: true,
        scale: 100.0,
        min_periods: 2,
        sigma_floor: 1e-6,
    };
    let kgz_local =
        cache.kalman_garch_zscore_local(Timeframe::M1, PriceType::Bid, 5, 4, kgz_params);
    assert_scalar_close(
        "kalman_garch_zscore_local",
        output.kalman_garch_zscore_local,
        kgz_local,
        1e-10,
    );

    let atr = ATR::new(3).compute(&primary_bid);
    assert_series_close("atr", &output.atr, &atr, 1e-10);

    let bb = BollingerBands::new(3, 2.0).compute_all(&primary_bid);
    assert_series_close("bollinger.upper", &output.bollinger.upper, &bb.upper, 1e-10);
    assert_series_close(
        "bollinger.middle",
        &output.bollinger.middle,
        &bb.middle,
        1e-10,
    );
    assert_series_close("bollinger.lower", &output.bollinger.lower, &bb.lower, 1e-10);

    let zscore_rolling = ZScore::new(3).compute(&primary_bid);
    assert_series_close(
        "zscore_rolling",
        &output.zscore_rolling,
        &zscore_rolling,
        1e-10,
    );

    let zscore_ema =
        ZScore::with_mean_source(3, ZScoreMeanSource::Ema, Some(2)).compute(&primary_bid);
    assert_series_close("zscore_ema", &output.zscore_ema, &zscore_ema, 1e-10);

    let kalman_zscore = KalmanZScore::new(3, 0.5, 0.1).compute(&primary_bid);
    assert_series_close(
        "kalman_zscore",
        &output.kalman_zscore,
        &kalman_zscore,
        1e-10,
    );

    let garch_vol = GarchVolatility::new(0.1, 0.8, 0.0)
        .with_log_returns(true)
        .with_scale(100.0)
        .with_min_periods(2)
        .with_sigma_floor(1e-6)
        .compute(&primary_bid);
    assert_series_close(
        "garch_volatility",
        &output.garch_volatility,
        &garch_vol,
        1e-10,
    );

    let kgz = KalmanGarchZScore::new(0.01, 1.0, 0.1, 0.8, 0.0)
        .with_log_returns(true)
        .with_scale(100.0)
        .with_min_periods(2)
        .with_sigma_floor(1e-6)
        .compute(&primary_bid);
    assert_series_close(
        "kalman_garch_zscore",
        &output.kalman_garch_zscore,
        &kgz,
        1e-10,
    );
}

fn make_candle(timestamp_ns: i64, close: f64) -> Candle {
    Candle {
        timestamp_ns,
        open: close - 0.01,
        high: close + 0.02,
        low: close - 0.02,
        close,
        volume: 1.0,
    }
}

fn assert_series_close(label: &str, expected: &[Option<f64>], actual: &[f64], atol: f64) {
    assert_eq!(
        expected.len(),
        actual.len(),
        "{label}: length mismatch {} != {}",
        expected.len(),
        actual.len()
    );

    for (idx, (exp, act)) in expected.iter().zip(actual.iter()).enumerate() {
        match exp {
            None => assert!(!act.is_finite(), "{label}[{idx}] expected NaN, got {act}"),
            Some(value) => {
                assert!(act.is_finite(), "{label}[{idx}] expected finite, got {act}");
                let diff = (value - act).abs();
                assert!(diff <= atol, "{label}[{idx}] diff {diff} exceeds {atol}");
            }
        }
    }
}

fn assert_scalar_close(label: &str, expected: Option<f64>, actual: Option<f64>, atol: f64) {
    match (expected, actual) {
        (None, None) => {}
        (Some(exp), Some(act)) => {
            let diff = (exp - act).abs();
            assert!(
                diff <= atol,
                "{label} diff {diff} exceeds {atol}: {exp} vs {act}"
            );
        }
        (Some(exp), None) => panic!("{label} expected {exp}, got None"),
        (None, Some(act)) => panic!("{label} expected None, got {act}"),
    }
}

fn run_python_parity() -> PythonParityOutput {
    let python = std::env::var("PYTHON").unwrap_or_else(|_| "python3".to_string());
    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let script_path = manifest_dir.join("tests").join("python_parity.py");

    let output = Command::new(&python)
        .arg(script_path)
        .output()
        .expect("Failed to run python parity script");

    if !output.status.success() {
        panic!(
            "Python parity script failed: {}",
            String::from_utf8_lossy(&output.stderr)
        );
    }

    serde_json::from_slice(&output.stdout).expect("Failed to parse python parity output")
}
